### PR TITLE
add some checks for fedration-apiserver options

### PR DESCRIPTION
**Pull request by m1093782566**
**What this PR does / why we need it**:

I find there is a TODO, see https://github.com/kubernetes/kubernetes/blob/master/federation/cmd/federation-apiserver/app/options/validation.go#L30

This PR add some checks for fedration-apiserver options

@sttts 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: federation/cmd/federation-apiserver/app/options/validation.go [hunk 1]
```diff
@@ -27,6 +29,18 @@ func (options *ServerRunOptions) Validate() []error {
...
+	if errs := options.Authentication.Validate(); len(errs) > 0 {
+		errors = append(errors, errs...)
+	}
+	if errs := options.Authorization.Validate(); len(errs) > 0 {
+		errors = append(errors, errs...)
+	}
+	if errs := options.CloudProvider.Validate(); len(errs) > 0 {
+		errors = append(errors, errs...)
+	}
+	if options.EventTTL <= 0 {
```
Comment (sttts) [inline federation/cmd/federation-apiserver/app/options/validation.go]: What about the other options?
```
	GenericServerRunOptions *genericoptions.ServerRunOptions
	Etcd                    *genericoptions.EtcdOptions
	SecureServing           *genericoptions.SecureServingOptions
	InsecureServing         *kubeoptions.InsecureServingOptions
	Audit                   *genericoptions.AuditOptions
	Features                *genericoptions.FeatureOptions
	Admission               *genericoptions.AdmissionOptions
	Authentication          *kubeoptions.BuiltInAuthenticationOptions
	Authorization           *kubeoptions.BuiltInAuthorizationOptions
	CloudProvider           *kubeoptions.CloudProviderOptions
	StorageSerialization    *kubeoptions.StorageSerializationOptions
	APIEnablement           *kubeoptions.APIEnablementOptions
```

Comment (m1093782566) [inline federation/cmd/federation-apiserver/app/options/validation.go]: I replied outline, please check. Thanks!

Comment (sttts) [inline federation/cmd/federation-apiserver/app/options/validation.go]: outline? Am confused :)

**Comment by m1093782566** [comment]
Glad to receive your reply, @sttts 

> What about the other options?

	GenericServerRunOptions *genericoptions.ServerRunOptions
	Etcd                    *genericoptions.EtcdOptions
	SecureServing           *genericoptions.SecureServingOptions
	InsecureServing         *kubeoptions.InsecureServingOptions
	Audit                   *genericoptions.AuditOptions
	Features                *genericoptions.FeatureOptions
	Admission               *genericoptions.AdmissionOptions
	Authentication          *kubeoptions.BuiltInAuthenticationOptions
	Authorization           *kubeoptions.BuiltInAuthorizationOptions
	CloudProvider           *kubeoptions.CloudProviderOptions
	StorageSerialization    *kubeoptions.StorageSerializationOptions
	APIEnablement           *kubeoptions.APIEnablementOptions

* GenericServerRunOptions

*genericoptions.ServerRunOptions has no `Validate()`, I am happy to implement it if it should have.

* Etcd

Already checked.

* SecureServing

Already checked.

* InsecureServing

Already checked.

* Audit

I just forgot to add the validation and now it's fixed.

* Features

 *genericoptions.FeatureOptions has no Validate() and the fields of it can be any bool type value. Should I add an empty Validate() for  *genericoptions.FeatureOptions?

* Admission

 *genericoptions.AdmissionOptions has no Validate() and the fields of it can be any string type value. Should I add an empty Validate() for  *genericoptions.FeatureOptions?

* Authentication

Checked.

* Authorization

Checked.

* CloudProvider

Checked.

* StorageSerialization

 *genericoptions.AdmissionOptions has no Validate() and the fields of it are validated in parser, see https://github.com/kubernetes/kubernetes/blob/master/pkg/kubeapiserver/options/storage_versions.go#L58

* APIEnablement

It's bool type and can be any bool type value.

* EventTTL

Checked.

File: federation/cmd/federation-apiserver/app/options/validation.go (hunk 1 above)
Comment (m1093782566) [inline federation/cmd/federation-apiserver/app/options/validation.go]: oops, I replied in outside context as the input box here is a bit small to reply.

**Comment by sttts** [comment]
@m1093782566 I had checked audit. So I was lucky :) Adding empty `Validate()` funcs where missing would be great. Makes more uniform.

**Comment by m1093782566** [comment]
> @m1093782566 I had checked audit. So I was lucky :) 

I think I am lucky as well :) 

Thanks for doing that.

> Adding empty Validate() funcs where missing would be great. Makes more uniform.

Sure, will do it soon.

**Comment by m1093782566** [comment]
@sttts 

All options are checked except  *genericoptions.ServerRunOptions. I guess validating *genericoptions.ServerRunOptions is out of scope of this PR and can we validate it in follow ups?

**Comment by sttts** [comment]
Looks good!

Can you also check:
* vendor/k8s.io/apiextensions-apiserver/pkg/cmd/server/start.go:80
* vendor/k8s.io/kube-aggregator/pkg/cmd/server/start.go:104
* vendor/k8s.io/sample-apiserver/pkg/cmd/server/start.go:82
* cmd/kube-apiserver/app/options/validation.go:49

Follow-up is fine.

**Comment by sttts** [comment]
/assign @nikhiljindal

Approved?

**Comment by m1093782566** [comment]
> Can you also check:
* vendor/k8s.io/apiextensions-apiserver/pkg/cmd/server/start.go:80
* vendor/k8s.io/kube-aggregator/pkg/cmd/server/start.go:104
* vendor/k8s.io/sample-apiserver/pkg/cmd/server/start.go:82
* cmd/kube-apiserver/app/options/validation.go:49

Of course, I am happy to do that. I created an issue for tracking them, please see: #50301. Thanks!

**Comment by m1093782566** [comment]
@quinton-hoole This PR needs your approval. Would you please take a look? Much appreciated!

**Comment by m1093782566** [comment]
Thanks all :)

**Comment by sttts** [comment]
@m1093782566 thanks for the contribution!

**Pull request merged**